python/vineyard/llm/tests/test_llm.py

```diff
@@ -32,71 +33,134 @@ def test_kv_cache_update_and_query_on_blob(vineyard_ipc_sockets):
     )
 
-    kv_cache_list = [
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-        (torch.rand(50, 50), torch.rand(50, 50)),
-    ]
-
     tokens = [1, 2, 3, 4]
-    # insert the token list and the related kv cache list
-    cache.update(tokens, kv_cache_list)
-
-    queried_kv_cache_list = cache.query(tokens)
+    original_kv_tensors = []
+    for i in range(0, len(tokens), file_cache_config.chunk_size):
+        kv_tensors_to_update = []
+        k_tensor = torch.rand(2, 2, dtype=torch.float32)
+        v_tensor = torch.rand(2, 2, dtype=torch.float32)
+        for _ in range(file_cache_config.chunk_size):
+            original_kv_tensors.append(
+                [(k_tensor, v_tensor) for _ in range(cache.layer)]
+            )
+            kv_tensors_to_update.append(
+                [
+                    (
+                        KVTensor(k_tensor.numpy().ctypes.data, k_tensor.nbytes),
+                        KVTensor(v_tensor.numpy().ctypes.data, v_tensor.nbytes),
+                    )
+                    for _ in range(cache.layer)
+                ]
+            )
+        updated = cache.update(
+            tokens[:i],
+            tokens[i : i + file_cache_config.chunk_size],
+            kv_tensors_to_update,
+        )
+        assert updated == file_cache_config.chunk_size
 
-    for (k_tensor, v_tensor), (queried_k_tensor, queried_v_tensor) in zip(
-        kv_cache_list, queried_kv_cache_list
-    ):
-        assert torch.equal(k_tensor, queried_k_tensor) and torch.equal(
-            v_tensor, queried_v_tensor
+    kv_tensors_from_cache = []
+    kv_tensors = []
+    for _ in range(len(tokens)):
+        k_tensor = torch.empty(2, 2, dtype=torch.float32)
+        v_tensor = torch.empty(2, 2, dtype=torch.float32)
+        kv_tensors_from_cache.append([(k_tensor, v_tensor) for _ in range(cache.layer)])
+        kv_tensors.append(
+            [
+                (
+                    KVTensor(k_tensor.numpy().ctypes.data, k_tensor.nbytes),
+                    KVTensor(v_tensor.numpy().ctypes.data, v_tensor.nbytes),
+                )
+                for _ in range(cache.layer)
+            ]
         )
+    matched = cache.query(tokens, kv_tensors)
+    assert matched == len(tokens)
+
+    assert len(kv_tensors) == len(kv_tensors_from_cache)
+    for kv, kv_from_cache in zip(original_kv_tensors, kv_tensors_from_cache):
+        assert len(kv) == len(kv_from_cache)
+        for (k_tensor, v_tensor), (queried_k_tensor, queried_v_tensor) in zip(
+            kv, kv_from_cache
+        ):
+            assert torch.equal(k_tensor, queried_k_tensor)
+            assert torch.equal(v_tensor, queried_v_tensor)
```

Comment:
Remove the usages of torch (as it is more harder to install). Numpy is enough for unittests of llm kvcache module.

Reply:
Ok.